internal/builtins/pkg_context.go

```diff
@@ -0,0 +1,130 @@
+// a KRM object that contains package context information that can be
+// used by functions such as `set-namespace` to customize package with
+// minimal configuration.
+type PackageContextGenerator struct{}
+
+func (pc *PackageContextGenerator) Run(r io.Reader, w io.Writer) error {
+	rw := &kio.ByteReadWriter{
+		Reader:                r,
+		Writer:                w,
+		KeepReaderAnnotations: true,
+	}
+	return framework.Execute(pc, rw)
+}
+
+func (pc *PackageContextGenerator) Process(resourceList *framework.ResourceList) error {
+	var contextResources, updatedResources []*yaml.RNode
+
+	// This loop does the following:
+	// - Filters out package context resources from the input resources
+	// - Generates a package context resource for each kpt package (i.e Kptfile)
+	for _, resource := range resourceList.Items {
+		if isPkgContext(resource) {
+			// drop existing package context resources
+			continue
+		}
+		updatedResources = append(updatedResources, resource)
+		if isKptfile(resource) {
+			pkgContext, err := pkgContextResource(resource)
+			if err != nil {
+				resourceList.Results = framework.Results{
+					&framework.Result{
+						Message:  err.Error(),
+						Severity: framework.Error,
+					},
+				}
+				return resourceList.Results
+			}
+			contextResources = append(contextResources, pkgContext)
+		}
+	}
+
+	for _, resource := range contextResources {
+		updatedResources = append(updatedResources, resource)
+		resourcePath, _, _ := kioutil.GetFileAnnotations(resource)
+		resourceList.Results = append(resourceList.Results, &framework.Result{
+			Message:  "generated package context",
+			Severity: framework.Info,
+			File:     &framework.File{Path: resourcePath, Index: 0},
+		})
+	}
+	resourceList.Items = updatedResources
+	return nil
+}
+
+// pkgContextResource generates package context resource from a given
+// Kptfile. The resource is generated adjacent to the Kptfile of the package.
+func pkgContextResource(kf *yaml.RNode) (*yaml.RNode, error) {
+	cm := yaml.MustParse(fmt.Sprintf(`
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: %s
+  annotations:
+    config.kubernetes.io/local-config: "true"
+data: {}
+`, pkgContextName))
+
+	kptfilePath, _, err := kioutil.GetFileAnnotations(kf)
+	if err != nil {
+		return nil, err
+	}
+	annotations := map[string]string{
+		kioutil.PathAnnotation: path.Join(path.Dir(kptfilePath), pkgContextFile),
+	}
+
+	for k, v := range annotations {
+		if _, err := cm.Pipe(yaml.SetAnnotation(k, v)); err != nil {
+			return nil, err
+		}
+	}
+	cm.SetDataMap(map[string]string{
+		"name": kf.GetName(),
+	})
+	return cm, nil
+}
+
+func isKptfile(resource *yaml.RNode) bool {
+	gvk := resid.GvkFromNode(resource)
+	return gvk.Kind == kptfilev1.KptFileName &&
+		gvk.ApiVersion() == kptfilev1.TypeMeta.APIVersion
+}
+
+func isPkgContext(resource *yaml.RNode) bool {
+	gvk := resid.GvkFromNode(resource)
+	return resource.GetName() == pkgContextName && gvk.Kind == "ConfigMap" && gvk.ApiVersion() == "v1"
```

Comment:
nit:
var pkgGVK = resid.NewGvk("kptfile.kpt.dev", "v1", "ConfigMap")

func isPkgContext() {
  gvk := resid.GvkFromNode(resource)
  return gvk.Equals(pkgGVK)
} # Or maybe don't need the func?